Perform MMS on GPUs

MMS needs to be performed on GPU. Only CPU MMS verification has been done

When using `amrex::ParallelFor` you'll have to create a standalone helper function that does the execution on GPU and call that function from the unit test because of how GTest creates tests within a `TEST_F` fixture. 

```console
[ 93%] Building CUDA object CMakeFiles/amr_wind_unit_tests.dir/unit_tests/mms/test_mms_error.cpp.o
/home/sanantha/exawind/source/amr-wind/unit_tests/mms/test_mms_error.cpp(38): error: The enclosing parent function ("TestBody") for an extended __device__ lambda cannot have private or protected access within its class

1 error detected in the compilation of "/scratch/sanantha/.tmp/tmpxft_0000351b_00000000-6_test_mms_error.cpp1.ii".
gmake[2]: *** [CMakeFiles/amr_wind_unit_tests.dir/unit_tests/mms/test_mms_error.cpp.o] Error 1
gmake[1]: *** [CMakeFiles/amr_wind_unit_tests.dir/all] Error 2
gmake: *** [all] Error 2
```

--- a/docs/sphinx_doc/dev/verification.rst
+++ b/docs/sphinx_doc/dev/verification.rst
@@ -3,7 +3,11 @@
 Verification
 ============
 
-Verification of AMR-WIND uses `MASA <https://github.com/manufactured-solutions/MASA>`_ and auto-differention tools to implement the Method of Manufactured Solutions.
+Verification of AMR-WIND uses `MASA
+<https://github.com/manufactured-solutions/MASA>`_ and
+auto-differention tools to implement the Method of Manufactured
+Solutions. MMS verification has been performed for both CPU and GPU
+architectures.
 
 Building and linking MASA
 ~~~~~~~~~~~~~~~~~~~~~~~~~

--- a/test/test_files/mms_godunov/mms_godunov.i
+++ b/test/test_files/mms_godunov/mms_godunov.i
@@ -2,7 +2,6 @@ time.max_step                = 10
 time.stop_time               = 5.0           # Max (simulated) time to evolve
 
 incflo.use_godunov                     = true
-incflo.godunov_include_diff_in_forcing = true
 
 time.fixed_dt = 0.05
 time.cfl              = 0.49           # CFL factor
@@ -32,13 +31,6 @@ incflo.physics = MMS
 MMS.masa_name = "navierstokes_3d_incompressible_homogeneous"
 ICNS.source_terms = MMSForcing
 
-incflo.gradrhoerr       = 0.1
-
-amr.plt_ccse_regtest    =  1
-amr.plt_vort            =  1
-amr.plt_p               =  1
-
-incflo.advect_tracer = true
 incflo.diffusion_type   = 2             # 0 = Explicit, 1 = Crank-Nicolson, 2 = Implicit
 
 incflo.verbose          =   1           # incflo_level

--- a/test/test_files/mms_godunov_plm/mms_godunov_plm.i
+++ b/test/test_files/mms_godunov_plm/mms_godunov_plm.i
@@ -3,7 +3,6 @@ time.stop_time               = 5.0           # Max (simulated) time to evolve
 
 incflo.use_godunov                     = true
 incflo.use_ppm                         = false
-incflo.godunov_include_diff_in_forcing = true
 
 time.fixed_dt = 0.05
 time.cfl              = 0.49           # CFL factor
@@ -33,13 +32,6 @@ incflo.physics = MMS
 MMS.masa_name = "navierstokes_3d_incompressible_homogeneous"
 ICNS.source_terms = MMSForcing
 
-incflo.gradrhoerr       = 0.1
-
-amr.plt_ccse_regtest    =  1
-amr.plt_vort            =  1
-amr.plt_p               =  1
-
-incflo.advect_tracer = true
 incflo.diffusion_type   = 2             # 0 = Explicit, 1 = Crank-Nicolson, 2 = Implicit
 
 incflo.verbose          =   1           # incflo_level

--- a/test/test_files/mms_mol/mms_mol.i
+++ b/test/test_files/mms_mol/mms_mol.i
@@ -2,7 +2,6 @@ time.max_step                = 10
 time.stop_time               = 5.0           # Max (simulated) time to evolve
 
 incflo.use_godunov                     = false
-incflo.godunov_include_diff_in_forcing = true
 
 time.fixed_dt = 0.05
 time.cfl              = 0.49           # CFL factor
@@ -32,13 +31,6 @@ incflo.physics = MMS
 MMS.masa_name = "navierstokes_3d_incompressible_homogeneous"
 ICNS.source_terms = MMSForcing
 
-incflo.gradrhoerr       = 0.1
-
-amr.plt_ccse_regtest    =  1
-amr.plt_vort            =  1
-amr.plt_p               =  1
-
-incflo.advect_tracer = true
 incflo.diffusion_type   = 2             # 0 = Explicit, 1 = Crank-Nicolson, 2 = Implicit
 
 incflo.verbose          =   1           # incflo_level

--- a/unit_tests/mms/test_mms_error.cpp
+++ b/unit_tests/mms/test_mms_error.cpp
@@ -7,6 +7,26 @@
 #include "MMS.H"
 
 namespace amr_wind_tests {
+
+namespace {
+void perturb_vel_field(
+    const amrex::Box& bx,
+    const amrex::Array4<amrex::Real>& vel,
+    const amrex::GpuArray<amrex::Real, AMREX_SPACEDIM>& dx,
+    const amrex::GpuArray<amrex::Real, AMREX_SPACEDIM>& problo)
+{
+    amrex::ParallelFor(
+        bx, [vel, dx, problo] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
+            const amrex::Real x = problo[0] + (i + 0.5) * dx[0];
+            const amrex::Real y = problo[1] + (j + 0.5) * dx[1];
+            const amrex::Real z = problo[2] + (k + 0.5) * dx[2];
+            vel(i, j, k, 0) += std::sin(amr_wind::utils::two_pi() * x);
+            vel(i, j, k, 1) += std::sin(amr_wind::utils::two_pi() * 2.0 * y);
+            vel(i, j, k, 2) += std::cos(amr_wind::utils::two_pi() * z);
+        });
+}
+} // namespace
+
 TEST_F(MMSMeshTest, mms_error)
 {
     populate_parameters();
@@ -34,16 +54,7 @@ TEST_F(MMSMeshTest, mms_error)
             const auto& dx = mesh().Geom(lev).CellSizeArray();
             const auto& problo = mesh().Geom(lev).ProbLoArray();
 
-            amrex::ParallelFor(
-                bx, [=] AMREX_GPU_DEVICE(int i, int j, int k) noexcept {
-                    const amrex::Real x = problo[0] + (i + 0.5) * dx[0];
-                    const amrex::Real y = problo[1] + (j + 0.5) * dx[1];
-                    const amrex::Real z = problo[2] + (k + 0.5) * dx[2];
-                    vel(i, j, k, 0) += std::sin(amr_wind::utils::two_pi() * x);
-                    vel(i, j, k, 1) +=
-                        std::sin(amr_wind::utils::two_pi() * 2.0 * y);
-                    vel(i, j, k, 2) += std::cos(amr_wind::utils::two_pi() * z);
-                });
+            perturb_vel_field(bx, vel, dx, problo);
         });
 
     const amrex::Real u_mms_err =

